### ISSUE TITLE
fatal: range without lower bound

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2558]
To reproduce:

```bash
PGRST_DB_MAX_ROWS=100 PGRST_DB_ANON_ROLE='postgrest_test_anonymous' postgrest-with-postgresql-14   postgrest-run
03/Nov/2022:17:53:37 -0500: Attempting to connect to the database...
03/Nov/2022:17:53:37 -0500: Connection successful
03/Nov/2022:17:53:37 -0500: Listening on port 3000
03/Nov/2022:17:53:37 -0500: Listening for notifications on the pgrst channel
03/Nov/2022:17:53:37 -0500: Config reloaded
03/Nov/2022:17:53:37 -0500: Schema cache loaded
FatalError {fatalErrorMessage = "range without lower bound"}
```

(Needs `max-rows`)

```
curl -X GET "localhost:3000/projects?limit=0"
Something went wrong
```

Bug introduced on https://github.com/PostgREST/postgrest/pull/2269.

The major issue is that every fatal error consumes a pool connection permanently, so if `PGRST_DB_POOL=1` is specified PostgREST becomes unresponsive.

**Comment by steve-chavez** [comment]
> The major issue is that every fatal error consumes a pool connection permanently, so if PGRST_DB_POOL=1 is specified PostgREST becomes unresponsive.

The above only happens on [v10.0.0.20221011](https://github.com/PostgREST/postgrest/releases/tag/v10.0.0.20221011) and [v10.1.0](https://github.com/PostgREST/postgrest/releases/tag/v10.1.0), likely because of an interaction with https://github.com/PostgREST/postgrest/pull/2449/. So it's only a major bug for those versions. We'd need to release a `v10.1.1` after this is fixed. 

On [v10.0.0](https://github.com/PostgREST/postgrest/releases/tag/v10.0.0) the same error happens but the pool is fine, we don't become unresponsive.

**Comment by steve-chavez** [comment]
The below line was alway a time bomb waiting to go off. We should remove it as part of this fix.

https://github.com/PostgREST/postgrest/blob/c8e4f38984d9194d939bf8b3c31010974f560be4/src/PostgREST/RangeQuery.hs#L63-L67

**Comment by steve-chavez** [comment]
> I wonder whether the idle_session_timeout mentioned https://github.com/PostgREST/postgrest/pull/2444#issuecomment-1300822796 would help with that? Even if PostgREST was reconnecting immediately - maybe that would still heal the connection?

Just tried it and nope. Somehow still unresponsive and it gets stuck at `Attempting to connect to the database...`

```
PGRST_DB_MAX_ROWS=100 PGRST_DB_POOL=1 PGRST_DB_ANON_ROLE='postgrest_test_anonymous' postgrest-with-postgresql-14   postgrest-run

04/Nov/2022:18:39:49 -0500: Attempting to connect to the database...
04/Nov/2022:18:39:49 -0500: Connection successful
04/Nov/2022:18:39:49 -0500: Listening on port 3000
04/Nov/2022:18:39:49 -0500: Listening for notifications on the pgrst channel
04/Nov/2022:18:39:49 -0500: Config reloaded
04/Nov/2022:18:39:49 -0500: Schema cache loaded
FatalError {fatalErrorMessage = "range without lower bound"}
FATAL:  terminating connection due to idle-session timeout
04/Nov/2022:18:39:54 -0500: Retrying listening for notifications on the pgrst channel..
04/Nov/2022:18:39:54 -0500: Attempting to connect to the database...

(No more mesages, retries or anything)
```

```
$ curl localhost:3000/tasks
.....
```

Don't know what's going on there.

**Comment by robx** [comment]
> The major issue is that every fatal error consumes a pool connection permanently, so if PGRST_DB_POOL=1 is specified PostgREST becomes unresponsive.

This sounds like an exception handling bug in `hasql-pool`. I'll have a look into that side of things.

**Comment by robx** [comment]
> > The major issue is that every fatal error consumes a pool connection permanently, so if PGRST_DB_POOL=1 is specified PostgREST becomes unresponsive.
> 
> This sounds like an exception handling bug in `hasql-pool`. I'll have a look into that side of things.

It *was* a but in hasql-pool, just fixed recently with release 0.8.0.5: https://github.com/nikita-volkov/hasql-pool/issues/25

**Comment by steve-chavez** [comment]
@wolfgangwalther WDYT of releasing a new patch release now(10.1.1)? Now that we have the new hasql-pool and the limit 0 fix I think removing the panic([ref](https://github.com/PostgREST/postgrest/issues/2558#issuecomment-1302851063)) is not critical and can wait.
